proposals/context.md

```diff
@@ -0,0 +1,265 @@
+  /**
+   * The name of the context that is requested
+   */
+  readonly name: T;
+  /**
+   * A boolean indicating if the context should only be provided once.
+   */
+  readonly once: boolean;
+  /**
+   * A callback which a provider of this named callback should invoke.
+   */
+  readonly callback: RequestContextCallback<ContextTypeMap[T]>;
+}
+```
+
+A full typescript definition for this event and its associated types can be found in the [Definitions](#definitions) section at the end of this document.
+
+## Context Providers
+
+A context provider will satisfy a `request-context` event, passing the `callback` the requested data whenever the data changes. A provider will attach an event listener to the DOM tree to catch the event, and if it will be able to satisfy the request will call `stopPropagation` on the event.
+
+If the provider has data available to satisfy the request then it should immediately invoke the `callback` passing the data. If the event does NOT have a truthy `once` property, then the provider can assume that the `callback` can be invokved multiple times, and may retain a reference to the callback to invoke as the data changes. If this is the case the provider should pass the second `dispose` parameter to the callback when invoking it in order to allow the requester to disconnect itself from the providers notifications.
+
+A provider does not necessarily have to be a Custom Element, but this may be a convenient mechanism.
+
+## Usage
+
+An element which wishes to receive some context and participate in the Context API should emit an event with the context-request type. This library includes an implementation of the ContextEvent class which implements the above interface and can be used in the following way:
+
+```js
+this.dispatchEvent(
+    new RequestContextEvent(
+        'cool-thing', // the name of the context we want to receive
+        callback: (coolThing) => {
+            this.myCoolThing = coolThing; // do something with value
+        }
+    )
+);
+```
+
+If a provider listening for this event can provide the requested context it will invoke the callback passed in the payload of the event. The element can then do whatever it wishes with this value.
+
+It may also be the case that a provider can retain a reference to this callback, and can then invoke the callback multiple times. In this case providers should pass a dispose function as a second argument to the callback to allow consumers to inform the provider that it should no longer update the element, and should dispose of the callback.
+
+As a convenience, and a hint for providers, an element may also provide a once boolean on the event detail to indicate that it is not interested in receiving updates to the value. If this behavior is essential to the correct operation of the consumer, then they should be implemented defensively as there is no guarantee that providers will honor this agreement. An example is provided below:
```

Comment:
It's more than just a hint, isn't it? If a consumer only wants a value once, then the provider doesn't have to keep the callback in memory or pass in a disconnect callback.

Along those lines, should `once` should default to true (or the option be `multiple: false`) in order to default to the lower-memory, less chance of a memory leak case?

Reply:
Agreed on the shift to `multiple` over `once`, have made this change in lit-context implementation, will update the proposal doc this afternoon.